funcx_endpoint/funcx_endpoint/executors/high_throughput/funcx_manager.py

```diff
@@ -420,18 +390,61 @@ def pull_tasks(self, kill_event):
-
-                            logger.debug("Sending task {} to {}".format(task.task_id, worker_id))
-                            # TODO: Some duplication of work could be avoided here
-                            to_send = [worker_id, pickle.dumps(task.task_id), pickle.dumps(task.container_id), task.pack()]
-                            self.funcx_task_socket.send_multipart(to_send)
-                            self.worker_map.update_worker_idle(task_type)
-                            if task.task_id != "KILL":
-                                logger.debug(f"Set task {task.task_id} to RUNNING")
-                                self.task_status_deltas[task.task_id] = TaskStatusCode.RUNNING
-                            logger.debug("Sending complete!")
+                            self.send_task_to_worker(task_type)
+
+    def poll_funcx_task_socket(self, task_done_counter):
+        try:
+            w_id, m_type, message = self.funcx_task_socket.recv_multipart()
+            if m_type == b'REGISTER':
+                reg_info = pickle.loads(message)
+                logger.debug("Registration received from worker:{} {}".format(w_id, reg_info))
+                self.worker_map.register_worker(w_id, reg_info['worker_type'])
+
+            elif m_type == b'TASK_RET':
+                logger.debug("Result received from worker: {}".format(w_id))
+                logger.debug("[TASK_PULL_THREAD] Got result: {}".format(message))
+                self.pending_result_queue.put(message)
+                self.worker_map.put_worker(w_id)
+                task_done_counter += 1
+                task_id = pickle.loads(message)['task_id']
+                task_type = self.task_type_mapping.pop(task_id)
+                self.task_status_deltas.pop(task_id, None)
+                logger.debug("Task type: {}".format(task_type))
+                self.outstanding_task_count[task_type] -= 1
+                logger.debug("Got result: Outstanding task counts: {}".format(self.outstanding_task_count))
+
+            elif m_type == b'WRKR_DIE':
+                logger.debug("[WORKER_REMOVE] Removing worker {} from worker_map...".format(w_id))
+                logger.debug("Ready worker counts: {}".format(self.worker_map.ready_worker_type_counts))
+                logger.debug("Total worker counts: {}".format(self.worker_map.total_worker_type_counts))
+                self.worker_map.remove_worker(w_id)
+                proc = self.worker_procs.pop(w_id.decode())
+                if not proc.poll():
+                    try:
+                        proc.wait(timeout=1)
+                    except subprocess.TimeoutExpired:
+                        logger.warning(f"[WORKER_REMOVE] Timeout waiting for worker {w_id} process to terminate")
+                logger.debug(f"[WORKER_REMOVE] Removing worker {w_id} process object")
+                logger.debug(f"[WORKER_REMOVE] Worker processes: {self.worker_procs}")
+
+            return pickle.loads(message), task_done_counter
```

Comment:
I worry that this is not safe. Do this instead: add a flag `poll_funcx_task_socket(test=False)` and based on the flag return   the pickle loaded message which is wrapped in try block.